apps/examples/webserver_test/webserver_test_main.c

```diff
@@ -0,0 +1,999 @@
+		ws_test_http_client_response_release(param->response);
+	}
+	if (param->tls) {
+		ws_test_tls_ssl_release(client_tls);
+	}
+	FUNC_EX;
+}
+
+int ws_test_parsehttpurl(const char *url, uint16_t *port, char *hostname, int hostlen) {
+	FUNC_EN;
+	const char *src = url;
+	char *dest;
+	int bytesleft;
+	int https = 0;
+
+	/* A valid HTTP URL must begin with http:// if it does not, we will assume
+	 * that it is a file name only, but still return an error.  wget() depends
+	 * on this strange behavior.
+	 */
+	if (strncmp(src, "https://", 8) == 0) {
+		https = 1;
+		*port = 443;
+	} else if (strncmp(src, "http://", 7) == 0) {
+		*port = 80;
+	} else {
+		return WS_TEST_ERR;
+	}
+
+	/* Skip over the http:// */
+	if (https) {
+		src += 8;
+	} else {
+		src += 7;
+	}
+
+	/* Concatenate the hostname following http:// and up to the termnator */
+	dest = hostname;
+	bytesleft = hostlen;
+	while (*src != '\0' && *src != '/' && *src != ' ' && *src != ':') {
+		/* Make sure that there is space for another character in the hostname.
+		 * (reserving space for the null terminator)
+		 */
+
+		if (bytesleft > 1) {
+			*dest++ = *src++;
+			bytesleft--;
+		} else {
+			return WS_TEST_ERR;
+		}
+	}
+	*dest = '\0';
+
+	FUNC_EX;
+	return WS_TEST_OK;
+}
+
+static int ws_test_base(void *arg)
+{
+	FUNC_EN;
+	int sockfd = -1;
+	int ret;
+	int buf_len, sndlen, len;
+	int encoding = HTTP_CONTENT_LENGTH;
+	int state = HTTP_REQUEST_HEADER;
+	struct http_message_len_t mlen = {0,};
+	struct ws_test_s ws;
+	struct http_client_request_t *param = (struct http_client_request_t *)arg;
+	struct http_client_response_t response = {0, };
+	int read_finish = false;
+	int send_len;
+
+
+	struct http_client_tls_t *client_tls = (struct http_client_tls_t *)malloc(sizeof(struct http_client_tls_t));
+	int handshake_retry = WS_TEST_CONF_HANDSHAKE_RETRY;
+
+	if (param->tls && client_tls == NULL) {
+		PRNT("memory not allocated");
+		return WS_TEST_ERR;
+	}
+
+	/* Initialize the state structure */
+	memset(&ws, 0, sizeof(struct ws_test_s));
+	ws.buffer = param->buffer;
+	ws.buflen = param->buflen;
+
+	/* Parse the hostname from the URL */
+	ret = ws_test_parsehttpurl(param->url, &ws.port, ws.hostname, WS_TEST_MAXHOSTNAME);
+	if (ret != 0) {
+		PRNT("ERROR: Malformed HTTP URL: %s", param->url);
+		free(client_tls);
+		return WS_TEST_ERR;
+	}
+
+	PRNT("hostname='%s'", ws.hostname);
+
+	if (param->tls && ws_test_tls_init(client_tls, &param->ssl_config)) {
+		PRNT("Fail to client tls init");
+		goto errout_before_tlsinit;
+	}
+
+	/* Re-initialize portions of the state structure that could have
+	 * been left from the previous time through the loop and should not
+	 * persist with the new connection.
+	 */
+	ws.httpstatus = HTTPSTATUS_NONE;
+	ws.offset = 0;
+
+	if ((sndlen = ws_test_msg_construct(ws.buffer, param, &ws)) < 0) {
+		PRNT("ERROR: construction message failed");
+		goto errout_before_tlsinit;
+	}
+
+	PRNT("sndlen:%d , param->buffer is:\n%s", sndlen,param->buffer);
+
+retry:
```

Comment:
using goto make code difficult to read. So how about use while() instead of goto.